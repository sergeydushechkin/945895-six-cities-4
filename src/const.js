const SortTypes = {
  POPULAR: `popular`,
  PRICE_LOW_HIGH: `priceLowHigh`,
  PRICE_HIGH_LOW: `priceHighLow`,
  TOP_RATED_FIRST: `topRatedFirst`
};

const SortTypeTexts = {
  [SortTypes.POPULAR]: `Popular`,
  [SortTypes.PRICE_LOW_HIGH]: `Price: low to high`,
  [SortTypes.PRICE_HIGH_LOW]: `Price: high to low`,
  [SortTypes.TOP_RATED_FIRST]: `Top rated first`,
};

export {SortTypes, SortTypeTexts};
