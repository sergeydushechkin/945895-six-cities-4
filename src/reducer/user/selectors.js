import NameSpace from "../name-space.js";

const NAME_SPACE = NameSpace.USER;

const getAuthStatus = (state) => {
  return state[NAME_SPACE].authorizationStatus;
};

const getAuthInfo = (state) => {
  return state[NAME_SPACE].authInfo;
};

export {getAuthStatus, getAuthInfo};
