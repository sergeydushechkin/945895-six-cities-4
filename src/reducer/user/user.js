import createAuthInfo from "../../adapters/auth-info/auth-info";
import {extend} from "../../utils";

const AuthorizationStatus = {
  AUTH: `AUTH`,
  NO_AUTH: `NO_AUTH`,
};

const initialState = {
  authorizationStatus: AuthorizationStatus.NO_AUTH,
  authInfo: {
    avatarUrl: ``,
    email: ``,
    id: null,
    isPro: null,
    name: ``,
  }
};

const ActionType = {
  CHANGE_AUTH_STATUS: `CHANGE_AUTH_STATUS`,
  CHANGE_AUTH_INFO: `CHANGE_AUTH_INFO`,
};

const ActionCreator = {
  changeAuthStatus: (authStatus) => ({
    type: ActionType.CHANGE_AUTH_STATUS,
    payload: authStatus
  }),
  changeAuthInfo: (authStatus) => ({
    type: ActionType.CHANGE_AUTH_INFO,
    payload: authStatus
  }),
};

const Operation = {
  checkAuth: () => (dispatch, getState, api) => {
    return api.get(`/login`)
      .then((response) => {
        dispatch(ActionCreator.changeAuthInfo(createAuthInfo(response.data)));
        dispatch(ActionCreator.changeAuthStatus(AuthorizationStatus.AUTH));
      })
      .catch((err) => {
        throw err;
      });
  },
  loginUser: (userData) => (dispatch, getState, api) => {
    return api.post(`/login`, userData)
      .then((response) => {
        dispatch(ActionCreator.changeAuthInfo(createAuthInfo(response.data)));
        dispatch(ActionCreator.changeAuthStatus(AuthorizationStatus.AUTH));
      })
      .catch((err) => {
        throw err;
      });
  }
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.CHANGE_AUTH_STATUS:
      return extend(state, {authorizationStatus: action.payload});
    case ActionType.CHANGE_AUTH_INFO:
      return extend(state, {authInfo: action.payload});
  }

  return state;
};

export {reducer, ActionCreator, ActionType, Operation, AuthorizationStatus};
