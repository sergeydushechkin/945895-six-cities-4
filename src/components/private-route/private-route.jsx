import React from "react";
import PropTypes from "prop-types";
import {Route, Redirect} from "react-router-dom";
import {connect} from "react-redux";

import {AuthorizationStatus} from "../../reducer/user/user.js";
import {getAuthStatus} from "../../reducer/user/selectors.js";

import {AppRoute} from "../../const.js";

const PrivateRoute = (props) => {
  const {render, path, exact, authStatus} = props;
  return (
    <Route
      path={path}
      exact={exact}
      render={() => {
        return (
          authStatus === AuthorizationStatus.AUTH
            ? render()
            : <Redirect to={AppRoute.LOGIN} />
        );
      }}
    />
  );
};

PrivateRoute.propTypes = {
  exact: PropTypes.bool.isRequired,
  path: PropTypes.string.isRequired,
  render: PropTypes.func.isRequired,
  authStatus: PropTypes.oneOf([AuthorizationStatus.AUTH, AuthorizationStatus.NO_AUTH]).isRequired,
};

const mapStateToProps = (state) => {
  return {
    authStatus: getAuthStatus(state),
  };
};

export {PrivateRoute};
export default connect(mapStateToProps)(PrivateRoute);
