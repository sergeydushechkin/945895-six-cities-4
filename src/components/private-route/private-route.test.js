import React from "react";
import renderer from "react-test-renderer";
import {Router} from "react-router-dom";

import {AuthorizationStatus} from "../../reducer/user/user.js";
import {AppRoute} from "../../const.js";
import history from "../../history.js";
import {PrivateRoute} from "./private-route.jsx";

it(`PrivateRoute render correct`, () => {
  const tree = renderer
    .create(
        <Router history={history}>
          <PrivateRoute
            exact
            path={AppRoute.FAVORITES}
            authStatus={AuthorizationStatus.AUTH}
            render={() => {
              return (<h1>Hello, World!</h1>);
            }}
          />
        </Router>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
