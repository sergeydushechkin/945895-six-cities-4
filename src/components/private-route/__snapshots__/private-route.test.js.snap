// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrivateRoute render correct 1`] = `null`;
